File: src/main/java/org/jukeboxmc/block/behavior/BlockLantern.java
package org.jukeboxmc.block.behavior;

import org.cloudburstmc.nbt.NbtMap;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;
import org.jukeboxmc.world.World;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockLantern extends Block {

    public BlockLantern( Identifier identifier ) {
        super( identifier );
    }

    public BlockLantern( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public boolean placeBlock( Player player, World world, Vector blockPosition, Vector placePosition, Vector clickedPosition, Item itemInHand, BlockFace blockFace ) {
        boolean hanging = blockFace != BlockFace.UP && this.isBlockAboveValid() && ( this.isBlockUnderValid() || blockFace == BlockFace.DOWN );
        if ( this.isBlockUnderValid() && !hanging ) {
            return false;
        }

        this.setHanging( hanging );
        return super.placeBlock( player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace );
    }

    private boolean isBlockAboveValid() {
        Block up = this.getSide( BlockFace.UP );
        if ( up instanceof BlockLeaves ) {
            return false;
        } else if ( up instanceof BlockFence || up instanceof BlockWall ) {
            return true;
        } else if ( up instanceof BlockSlab ) {
            return !( (BlockSlab) up ).isTopSlot();
        } else if ( up instanceof BlockStairs ) {
            return !( (BlockStairs) up ).isUpsideDown();
        } else return up.isSolid();
    }

    private boolean isBlockUnderValid() {
        Block down = this.getSide( BlockFace.DOWN );
        if ( down instanceof BlockLeaves ) {
            return true;
        } else if ( down instanceof BlockFence || down instanceof BlockWall ) {
            return false;
        } else if ( down instanceof BlockSlab ) {
            return !( (BlockSlab) down ).isTopSlot();
        } else if ( down instanceof BlockStairs ) {
            return !( (BlockStairs) down ).isUpsideDown();
        } else return !down.isSolid();
    }

    public void setHanging( boolean value ) {
        this.setState( "hanging", value ? 1 : 0 );
    }

    public boolean isHanging() {
        return this.stateExists( "hanging" ) && this.getByteState( "hanging" ) == 1;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockLantern.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Lantern
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.block.data.VerticalHalf
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.extensions.toByte
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockLantern(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates), Lantern {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        val hanging =
            blockFace != BlockFace.UP && this.isBlockAboveValid() && (this.isBlockUnderValid() || blockFace == BlockFace.DOWN)
        if (this.isBlockUnderValid() && !hanging) {
            return false
        }
        val block = world.getBlock(placePosition)
        if (block is BlockWater && block.getLiquidDepth() == 0) {
            world.setBlock(placePosition, block, 1, false)
        }
        this.setHanging(hanging)
        return super.placeBlock(player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace)
    }

    private fun isBlockAboveValid(): Boolean {
        val up = this.getRelative(BlockFace.UP)
        return if (up.getType().isLeaves()) {
            false
        } else if (up.getType().isFence() || up.getType().isWall()) {
            true
        } else if (up is BlockSlab) {
            up.getVerticalHalf() != VerticalHalf.TOP
        } else if (up is BlockStairs) {
            !up.isUpsideDown()
        } else up.isSolid()
    }

    private fun isBlockUnderValid(): Boolean {
        val down = this.getRelative(BlockFace.DOWN)
        return if (down.getType().isLeaves()) {
            true
        } else if (down.getType().isFence() || down.getType().isWall()) {
            false
        } else if (down is BlockSlab) {
            down.getVerticalHalf() != VerticalHalf.TOP
        } else if (down is BlockStairs) {
            !down.isUpsideDown()
        } else !down.isSolid()
    }

    override fun isHanging(): Boolean {
        return this.getBooleanState("hanging")
    }

    override fun setHanging(value: Boolean): Lantern {
        return this.setState("hanging", value.toByte())
    }
}